### PR TITLE
Feat/v4 contracts

### PR DESCRIPTION
**Pull request by Dozie2001**
## Description
This PR refactors the semaphore extension contracts to use the v4 [`@semaphore-protocol/contracts`](https://github.com/semaphore-protocol/semaphore/tree/v3.15.2/packages/contracts/contracts/extensions) and moves them to the [extensions repo](https://github.com/semaphore-protocol/extensions) 

It removes the need for tree depth in Groups

## Related Issue(s)
Closes [#353](https://github.com/semaphore-protocol/semaphore/issues/353)

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [ ] My changes generate no new warnings
-   [x] I have run `yarn style` without getting any errors
-   [x] I have added tests that prove my fix is effective or that my feature works
-   [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/contracts/.yarn/releases/yarn-4.1.0.cjs
Comment (cedoor) [inline packages/contracts/.yarn/releases/yarn-4.1.0.cjs:1]: We don't actually need this here.

So, you can remove the following files/folders from this package: `.yarn`, `.yarnrc.yml`, `.editorconfig`, `.eslintignore`, `.eslint.json`, `.prettierignore`, `.gitignore`.

File: packages/contracts/.gitignore [hunk 1]
```diff
@@ -0,0 +1,71 @@
+node_modules
+.env
+
+# Hardhat files
+/cache
+/artifacts
```
Comment (cedoor) [inline packages/contracts/.gitignore:6]: These lines can be moved to the `.gitignore` file in the root folder and this file can be removed.

File: packages/contracts/README.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+```bash
+yarn prettier
+```
+
+Or to automatically format the code:
+
+```bash
+yarn prettier:write
+```
```
Comment (cedoor) [inline packages/contracts/README.md:108]: This section (Code quality and formatting) is not necessary here. Those commands are supposed to be executed in the root folder.

File: packages/contracts/README.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    <h1 align="center">
+        Semaphore Extension Contracts
+    </h1>
+    <p align="center">Extension contracts to manage Voting and whistleBlowing</p>
+</p>
+
+<p align="center">
+    <a href="https://github.com/semaphore-protocol">
+        <img src="https://img.shields.io/badge/project-Semaphore-blue.svg?style=flat-square">
+    </a>
```
Comment (cedoor) [inline packages/contracts/README.md:11]: Missing some tags here:

```html
<p align="center">
    <a href="https://github.com/semaphore-protocol">
        <img src="https://img.shields.io/badge/project-Semaphore-blue.svg?style=flat-square">
    </a>
    <a href="https://github.com/semaphore-protocol/extensions/blob/main/packages/contracts/contracts/LICENSE">
        <img alt="NPM license" src="https://img.shields.io/npm/l/%40semaphore-extensions%2Fcontracts?style=flat-square">
    </a>
    <a href="https://www.npmjs.com/package/@semaphore-extensions/contracts">
        <img alt="NPM version" src="https://img.shields.io/npm/v/@semaphore-extensions/contracts?style=flat-square" />
    </a>
    <a href="https://npmjs.org/package/@semaphore-extensions/contracts">
        <img alt="Downloads" src="https://img.shields.io/npm/dm/@semaphore-extensions/contracts.svg?style=flat-square" />
    </a>
</p>
```

File: packages/contracts/README.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+## 🛠 Install
+
+### npm or yarn
+
+Install the `@semaphore-extension/contracts` package with npm:
+
+```bash
+npm i @semaphore-extension/contracts
+```
+
```
Comment (cedoor) [inline packages/contracts/README.md:41]: Please add an example with Yarn too.

```
or yarn:

```bash
yarn add @semaphore-protocol/contracts
\```
```

File: packages/contracts/README.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+2. Add your environment variables.
+
+> **Note**  
+> You should at least set a valid Ethereum URL (e.g. Infura) and a private key with some ethers.
+
+### Deploy contracts
+
+Deploy the `Semaphore.sol` contract without any parameter:
```
Comment (cedoor) [inline packages/contracts/README.md:81]: This section needs to be updated.

File: packages/contracts/README.md
Comment (cedoor) [inline packages/contracts/README.md:1]: This file should actually be a symbolic link of the readme file in `contracts/contracts/README.md`.

File: packages/contracts/contracts/package.json [hunk 6]
```diff
@@ -0,0 +1,19 @@
+{
+    "name": "@semaphore-extensions/contracts",
+    "version": "4.0.0-beta.16",
```
Comment (cedoor) [inline packages/contracts/contracts/package.json:3]: The version can be 0.1.0 here

File: packages/contracts/contracts/package.json [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+        "README.md"
+    ],
+    "homepage": "https://github.com/semaphore-protocol/extensions",
+    "publishConfig": {
+        "access": "public"
+    },
+    "dependencies": {
+        "@semaphore-protocol/contracts": "^4.0.0-beta.16"
+    },
+    "packageManager": "yarn@4.1.0"
```
Comment (cedoor) [inline packages/contracts/contracts/package.json:18]: No need to specify the package manager here

File: packages/contracts/package.json [hunk 8]
```diff
@@ -0,0 +1,79 @@
+{
+    "name": "extension-contracts",
+    "version": "4.0.0-beta.16",
```
Comment (cedoor) [inline packages/contracts/package.json:3]: No need to specify any version here.

File: packages/contracts/package.json [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        ".eslintrc.json",
+        ".prettierignore",
+        ".prettierrc.json",
+        "hardhat.config.ts",
+        "tsconfig.json",
+        "README.md"
+    ],
+    "publishConfig": {
+        "access": "public"
+    },
```
Comment (cedoor) [inline packages/contracts/package.json:22]: No need to specify these fields here.

File: packages/contracts/package.json [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        "hardhat-gas-reporter": "^1.0.8",
+        "prettier": "^3.2.5",
+        "prettier-plugin-solidity": "^1.3.1",
+        "solhint": "^4.1.1",
+        "solidity-coverage": "^0.8.1",
+        "ts-node": "^10.9.2",
+        "typechain": "^8.3.0",
+        "typescript": "^5.3.3"
+    },
+    "packageManager": "yarn@4.1.0",
```
Comment (cedoor) [inline packages/contracts/package.json:75]: No need to specify the package manager here.

**Review by cedoor — CHANGES_REQUESTED** [review]
Hey @Dozie2001, thanks for this PR!

First round of review. I haven't seen the code yet but let's start with this config stuff.

File: packages/contracts/README.md
Comment (Dozie2001) [reply packages/contracts/README.md:1]: What do you mean by this?

Comment (cedoor) [reply packages/contracts/README.md:1]: Take a look here: https://github.com/semaphore-protocol/semaphore/tree/main/packages/contracts

The README.md file is not a normal file. It's a symbolic link created with `ln -s ...`. The real README.md file is in the `packages/contracts/contracts` folder.

https://linuxhandbook.com/symbolic-link-linux

File: packages/contracts/contracts/README.md [hunk 11]
```diff
@@ -0,0 +1,103 @@
+<p align="center">
+    <h1 align="center">
+        Semaphore Extension Contracts
+    </h1>
+    <p align="center">Extension contracts to manage Voting and whistleBlowing</p>
```
Comment (cedoor) [inline packages/contracts/contracts/README.md:5]: The description should be more general, as there will be other extensions.

Something like "Solidity contracts to enhance Semaphore with extra features and customization."

File: packages/contracts/contracts/README.md [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+### npm or yarn
+
+Install the `@semaphore-extension/contracts` package with npm:
+
+```bash
+npm i @semaphore-extension/contracts
+```
+
+or yarn: <br>
```
Comment (cedoor) [inline packages/contracts/contracts/README.md:51]: I don't think we need `<br>` here.

File: packages/contracts/contracts/package.json [hunk 13]
```diff
@@ -0,0 +1,18 @@
+{
+    "name": "@semaphore-extensions/contracts",
+    "version": "0.1.0",
+    "description": "Semaphore extensions Contracts to manage anonymous voting and whistleblowing",
```
Comment (cedoor) [inline packages/contracts/contracts/package.json:4]: Remember to update this description based on the comment in the README file above.

File: packages/contracts/package.json [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+    "description": "Semaphore extensions Contracts to manage anonymous voting and whistleblowing",
+    "publishConfig": {
+        "access": "public"
+    },
+    "scripts": {
+        "dev": "hardhat node & yarn deploy --network localhost",
+        "compile": "hardhat compile",
+        "deploy": "yarn deploy:semaphore-voting && yarn deploy:semaphore-whistleblowing",
+        "deploy:semaphore-voting": "yarn compile && hardhat deploy:semaphore-voting",
+        "deploy:semaphore-whistleblowing": "hardhat compile && hardhat deploy:semaphore-whistleblowing",
```
Comment (cedoor) [inline packages/contracts/package.json:12]: I don't think it's necessary to compile contracts before deploying here.

File: packages/contracts/package.json [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+        "verify": "hardhat verify",
+        "test": "hardhat test",
+        "test:report-gas": "REPORT_GAS=true hardhat test",
+        "test:coverage": "hardhat coverage",
+        "typechain": "hardhat typechain",
+        "prepublish": "tar -czf files.tgz .gitignore .yarn .yarnrc.yml",
+        "lint": "eslint . --ext .js,.ts && solhint 'contracts/**/*.sol'",
```
Comment (cedoor) [inline packages/contracts/package.json:19]:
```suggestion
        "lint": "solhint 'contracts/**/*.sol'",
```

eslint needs to be run in the root folder

File: packages/contracts/package.json [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+        "prettier": "prettier -c --plugin=prettier-plugin-solidity .",
+        "prettier:write": "prettier -w --plugin=prettier-plugin-solidity ."
```
Comment (cedoor) [inline packages/contracts/package.json:21]: Prettier is supposed to be run in the root folder.

File: packages/contracts/package.json [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+        "compile": "hardhat compile",
+        "deploy": "yarn deploy:semaphore-voting && yarn deploy:semaphore-whistleblowing",
+        "deploy:semaphore-voting": "yarn compile && hardhat deploy:semaphore-voting",
+        "deploy:semaphore-whistleblowing": "hardhat compile && hardhat deploy:semaphore-whistleblowing",
+        "verify": "hardhat verify",
+        "test": "hardhat test",
+        "test:report-gas": "REPORT_GAS=true hardhat test",
+        "test:coverage": "hardhat coverage",
+        "typechain": "hardhat typechain",
+        "prepublish": "tar -czf files.tgz .gitignore .yarn .yarnrc.yml",
```
Comment (cedoor) [inline packages/contracts/package.json:18]: Not sure why this script is here. It doesn't look necessary.

File: packages/contracts/contracts/SemaphoreVoting.sol [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+
+import {ISemaphore} from "@semaphore-protocol/contracts/interfaces/ISemaphore.sol";
+import {ISemaphoreVoting} from "./interfaces/ISemaphoreVoting.sol";
+import {SemaphoreGroups} from "@semaphore-protocol/contracts/base/SemaphoreGroups.sol";
+
+/// @title Semaphore Voting contract.
+/// @notice It allows users to vote anonymously in a poll.
+/// @dev The following code allows you to create polls, add voters and allow them to vote anonymously.
+contract SemaphoreVoting is ISemaphoreVoting, SemaphoreGroups {
+    ISemaphore public group;
```
Comment (cedoor) [inline packages/contracts/contracts/SemaphoreVoting.sol:12]:
```suggestion
    ISemaphore public semaphore;
```

This is not the group, it's the semaphore contract. Please update all the occurrences and the comments below.

File: packages/contracts/contracts/SemaphoreVoting.sol [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+
+    /// @dev See {ISemaphoreVoting-castVote}.
+    function castVote(uint256 vote, uint256 nullifierHash, uint256 pollId, uint256[8] calldata proof) public override {
+        if (polls[pollId].state != PollState.Ongoing) {
+            revert Semaphore__PollIsNotOngoing();
+        }
+
+        if (polls[pollId].nullifierHashes[nullifierHash]) {
+            revert Semaphore__YouAreUsingTheSameNullifierTwice();
+        }
```
Comment (cedoor) [inline packages/contracts/contracts/SemaphoreVoting.sol:71]: This is not necessary, the `Semaphore` contract already checks it.

File: packages/contracts/contracts/SemaphoreVoting.sol [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+
+        polls[pollId].state = PollState.Ongoing;
+
+        emit PollStarted(pollId, msg.sender, encryptionKey);
+    }
+
+    /// @dev See {ISemaphoreVoting-castVote}.
+    function castVote(uint256 vote, uint256 nullifierHash, uint256 pollId, uint256[8] calldata proof) public override {
+        if (polls[pollId].state != PollState.Ongoing) {
+            revert Semaphore__PollIsNotOngoing();
```
Comment (cedoor) [inline packages/contracts/contracts/SemaphoreVoting.sol:66]: Good practice for error names: <contractName>__<ErrorName>. Could you replace Semaphore with the name of this contract?

Remember to update the other error names as well.

File: packages/contracts/contracts/SemaphoreVoting.sol [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    /// @dev Initializes the Semaphore group used to verify the user's ZK proofs.
+    /// @param _group: Semaphore group address.
+    constructor(ISemaphore _group) {
+        group = _group;
+    }
+
+    /// @dev See {ISemaphoreVoting-createPoll}.
+    function createPoll(uint256 pollId, address coordinator) public override {
```
Comment (cedoor) [inline packages/contracts/contracts/SemaphoreVoting.sol:34]: We can get rid of the `pollId` parameter here. We can use the `groupId` returned by `createGroup()` directly.

File: packages/contracts/contracts/SemaphoreVoting.sol [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+            merkleTreeRoot: merkleTreeRoot,
+            nullifier: nullifierHash,
+            message: vote,
+            scope: pollId,
+            points: proof
+        });
+
+        group.verifyProof(polls[pollId].groupId, semaphoreProof);
+
+        polls[pollId].nullifierHashes[nullifierHash] = true;
```
Comment (cedoor) [inline packages/contracts/contracts/SemaphoreVoting.sol:86]: Not necessary.

**Review by cedoor — CHANGES_REQUESTED** [review]
A few comments on config stuff and review on Voting contract.

File: packages/contracts/contracts/SemaphoreVoting.sol (hunk 20 above)
Comment (Dozie2001) [reply packages/contracts/contracts/SemaphoreVoting.sol:66]: don't quite understand this comment please.

File: packages/contracts/.eslintrc.json
Comment (cedoor) [inline packages/contracts/.eslintrc.json:1]: Could you please remove this file?

File: packages/contracts/contracts/README.md [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+
+```bash
+yarn deploy
+```
+
+> **Note**  
+> You should at least set a valid Ethereum URL (e.g. Infura) and a private key with some ethers.
+
+> **Warning**  
+> The group id is a number!
```
Comment (cedoor) [inline packages/contracts/contracts/README.md:103]: This documentation needs to be updated.

File: packages/contracts/contracts/SemaphoreVoting.sol [hunk 24]
```diff
@@ -0,0 +1,88 @@
...
+
+    /// @dev See {ISemaphoreVoting-createPoll}.
+    function createPoll(address coordinator) public {
+        uint256 groupId = semaphore.createGroup();
+
+        emit PollCreated(groupId, coordinator);
+    }
+
+    /// @dev See {ISemaphoreVoting-addVoter}.
+    function addVoter(uint256 pollId, uint256 groupId, uint256 identityCommitment) public onlyCoordinator(pollId) {
```
Comment (cedoor) [inline packages/contracts/contracts/SemaphoreVoting.sol:41]: pollId and groupId should be the same value here. So we can remove groupId

File: packages/contracts/contracts/SemaphoreVoting.sol [hunk 25]
```diff
@@ -0,0 +1,88 @@
...
+
+    function addVoter(uint256 pollId, uint256 identityCommitment) external override {}
+
+    function castVote(
+        uint256 vote,
+        uint256 nullifierHash,
+        uint256 pollId,
+        uint256[8] calldata proof
+    ) external override {}
+}
```
Comment (cedoor) [inline packages/contracts/contracts/SemaphoreVoting.sol:88]: I'm not sure why we have these unimplemented functions here.

File: packages/contracts/contracts/SemaphoreVoting.sol [hunk 26]
```diff
@@ -0,0 +1,88 @@
...
+
+    /// @dev Initializes the Semaphore group used to verify the user's ZK proofs.
+    /// @param _semaphore: Semaphore group address.
+    constructor(ISemaphore _semaphore) {
+        semaphore = _semaphore;
+    }
+
+    /// @dev See {ISemaphoreVoting-createPoll}.
+    function createPoll(address coordinator) public {
+        uint256 groupId = semaphore.createGroup();
```
Comment (cedoor) [inline packages/contracts/contracts/SemaphoreVoting.sol:35]: This should add a new item to the `polls` mapping.

File: packages/contracts/contracts/SemaphoreWhistleblowing.sol [hunk 27]
```diff
@@ -0,0 +1,120 @@
...
+
+    function updateGroupAdmin(uint256 groupId, address newAdmin) external override {}
+
+    function acceptGroupAdmin(uint256 groupId) external override {}
+
+    function updateGroupMerkleTreeDuration(uint256 groupId, uint256 newMerkleTreeDuration) external override {}
+
+    function addMember(uint256 groupId, uint256 identityCommitment) external override {}
+
+    function addMembers(uint256 groupId, uint256[] calldata identityCommitments) external override {}
```
Comment (cedoor) [inline packages/contracts/contracts/SemaphoreWhistleblowing.sol:93]: What is this list of override functions?

File: packages/contracts/contracts/SemaphoreWhistleblowing.sol [hunk 28]
```diff
@@ -0,0 +1,120 @@
...
+        uint256 identityCommitment,
+        uint256[] calldata proofSiblings
+    ) external onlyEditor(entityId) {
+        semaphore.removeMember(entityId, identityCommitment, proofSiblings);
+    }
+
+    /// @dev See {ISemaphoreWhistleblowing-publishLeak}.
+    function publishLeak(
+        uint256 leak,
+        uint256 nullifierHash,
```
Comment (cedoor) [inline packages/contracts/contracts/SemaphoreWhistleblowing.sol:57]: Please rename nullifierHash to nullifier, which is the new name we're using on Semaphore v4.

**Review by cedoor — CHANGES_REQUESTED** [review]

File: packages/contracts/contracts/SemaphoreWhistleblowing.sol (hunk 27 above)
Comment (Dozie2001) [reply packages/contracts/contracts/SemaphoreWhistleblowing.sol:93]: for some reason, the contract adds them unless it becomes an abstract contract.

File: packages/contracts/contracts/README.md [hunk 29]
```diff
@@ -96,8 +96,5 @@ Deploy the `SemaphoreVoting.sol` and `SemaphoreWhistleblowing.sol` contract with
 yarn deploy
 ```
 
-> **Note**  
-> You should at least set a valid Ethereum URL (e.g. Infura) and a private key with some ethers.
-
```
Comment (cedoor) [inline packages/contracts/contracts/README.md]: This was actually right, the warning about the group id below is wrong.

File: packages/contracts/contracts/SemaphoreWhistleblowing.sol [hunk 30]
```diff
@@ -8,7 +8,7 @@ import {SemaphoreGroups} from "@semaphore-protocol/contracts/base/SemaphoreGroup
 /// @title SemaphoreWhistleblowing
 /// @dev This contract uses the Semaphore base contracts to allow whistleblowers to leak information anonymously
 /// Leaks can be IPFS hashes, permanent links or other kinds of references.
-contract SemaphoreWhistleblowing is ISemaphoreWhistleblowing, SemaphoreGroups {
+abstract contract SemaphoreWhistleblowing is ISemaphoreWhistleblowing, SemaphoreGroups {
```
Comment (cedoor) [inline packages/contracts/contracts/SemaphoreWhistleblowing.sol:11]: It shouldn't be an abstract contract. And it shouldn't extend `SemaphoreGroups`.

**Review by cedoor — CHANGES_REQUESTED** [review]

File: packages/contracts/contracts/SemaphoreWhistleblowing.sol (hunk 27 above)
Comment (cedoor) [reply packages/contracts/contracts/SemaphoreWhistleblowing.sol]: I think you need to remove `SemaphoreGroups` from the extended contracts.

File: packages/contracts/contracts/SemaphoreWhistleblowing.sol (hunk 30 above)
Comment (Dozie2001) [reply packages/contracts/contracts/SemaphoreWhistleblowing.sol]: Hey I fixed this but havin some issue using the ```getmerkleTreeDuration```

File: packages/contracts/package.json (hunk 10 above)
Comment (cedoor) [reply packages/contracts/package.json:53]: @Dozie2001 could you remove the packageManager field here?

File: packages/contracts/contracts/package.json (hunk 13 above)
Comment (cedoor) [reply packages/contracts/contracts/package.json:4]: Please remember to update this description.

File: packages/contracts/contracts/SemaphoreWhistleblowing.sol [hunk 31]
```diff
@@ -0,0 +1,75 @@
...
+    function publishLeak(
+        uint256 leak,
+        uint256 nullifier,
+        uint256 entityId,
+        uint256[8] calldata proof
+    ) external override {
+        uint256 merkleTreeRoot = getmerkleTreeDuration(entityId);
+
+        ISemaphore.SemaphoreProof memory semaphoreProof = ISemaphore.SemaphoreProof({
+            merkleTreeDepth: 32,
```
Comment (cedoor) [inline packages/contracts/contracts/SemaphoreWhistleblowing.sol:63]: The tree depth should be passed as a parameter.

File: packages/contracts/contracts/SemaphoreWhistleblowing.sol [hunk 32]
```diff
@@ -0,0 +1,75 @@
...
+            merkleTreeRoot: merkleTreeRoot,
+            nullifier: nullifier,
+            message: leak,
+            scope: entityId,
+            points: proof
+        });
+
+        semaphore.verifyProof(entityId, semaphoreProof);
```
Comment (cedoor) [inline packages/contracts/contracts/SemaphoreWhistleblowing.sol:71]: You should use `validateProof` here.

File: packages/contracts/contracts/SemaphoreWhistleblowing.sol [hunk 33]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    /// @dev See {ISemaphoreWhistleblowing-publishLeak}.
+    function publishLeak(
+        uint256 leak,
+        uint256 nullifier,
+        uint256 entityId,
+        uint256[8] calldata proof
+    ) external override {
+        uint256 merkleTreeRoot = getmerkleTreeDuration(entityId);
```
Comment (cedoor) [inline packages/contracts/contracts/SemaphoreWhistleblowing.sol:60]: Re https://github.com/semaphore-protocol/extensions/pull/14#discussion_r1702139981

The merkle root should also be passed as a parameter here.

File: packages/contracts/contracts/SemaphoreVoting.sol [hunk 34]
```diff
@@ -0,0 +1,77 @@
...
+            revert SemaphoreVoting__PollHasAlreadyBeenStarted();
+        }
+
+        polls[pollId].state = PollState.Ongoing;
+
+        emit PollStarted(pollId, msg.sender, encryptionKey);
+    }
+
+    /// @dev See {ISemaphoreVoting-castVote}.
+    function castVote(uint256 vote, uint256 pollId) public {
```
Comment (cedoor) [inline packages/contracts/contracts/SemaphoreVoting.sol:63]: This function should validate the proof, like `publishLeak`.

**Review by cedoor — CHANGES_REQUESTED** [review]

File: packages/contracts/contracts/SemaphoreWhistleblowing.sol (hunk 30 above)
Comment (cedoor) [reply packages/contracts/contracts/SemaphoreWhistleblowing.sol]: -> https://github.com/semaphore-protocol/extensions/pull/14#discussion_r1703835738

File: packages/contracts/contracts/README.md [hunk 35]
```diff
@@ -0,0 +1,100 @@
...
+
+> **Note**  
+> You should at least set a valid Ethereum URL (e.g. Infura) and a private key with some ethers.
+
+### Deploy contracts
+
+Deploy the `SemaphoreVoting.sol` and `SemaphoreWhistleblowing.sol` contract without any parameter:
+
+```bash
+yarn deploy
```
Comment (cedoor) [inline packages/contracts/contracts/README.md:99]: Please update this command. There should be two: `deploy:semaphore-voting` and `deploy:semaphore-whistleblowing`

File: packages/contracts/contracts/SemaphoreVoting.sol [hunk 36]
```diff
@@ -0,0 +1,95 @@
...
+            merkleTreeDepth: merkleTreeDepth,
+            merkleTreeRoot: merkleTreeRoot,
+            nullifier: nullifier,
+            message: vote,
+            scope: pollId,
+            points: proof
+        });
+
+        semaphore.validateProof(pollId, semaphoreProof);
+        polls[pollId].nullifiers[nullifier] = true;
```
Comment (cedoor) [inline packages/contracts/contracts/SemaphoreVoting.sol:85]: This is not necessary, the `Semaphore` contract is already storing the nullifiers.

File: packages/contracts/contracts/interfaces/ISemaphoreVoting.sol [hunk 37]
```diff
@@ -0,0 +1,82 @@
...
+    enum PollState {
+        Created,
+        Ongoing,
+        Ended
+    }
+
+    struct Verifier {
+        address contractAddress;
+        uint256 merkleTreeDepth;
+    }
```
Comment (cedoor) [inline packages/contracts/contracts/interfaces/ISemaphoreVoting.sol:21]: This is unused right? We can remove it I guess.

File: packages/contracts/contracts/SemaphoreWhistleblowing.sol [hunk 38]
```diff
@@ -0,0 +1,82 @@
...
+
+        emit LeakPublished(entityId, leak);
+    }
+
+    function removeWhistleblower(
+        uint256 entityId,
+        uint256 identityCommitment,
+        uint256[] calldata proofSiblings,
+        uint8[] calldata proofPathIndices
+    ) external override {}
```
Comment (cedoor) [inline packages/contracts/contracts/SemaphoreWhistleblowing.sol:81]: This code should be removed.

File: packages/contracts/contracts/package.json [hunk 39]
```diff
@@ -0,0 +1,18 @@
...
+        "**/*.sol",
+        "*.sol",
+        "README.md"
+    ],
+    "homepage": "https://github.com/semaphore-protocol/extensions",
+    "publishConfig": {
+        "access": "public"
+    },
+    "dependencies": {
+        "@semaphore-protocol/contracts": "^4.0.0-beta.16"
```
Comment (cedoor) [inline packages/contracts/contracts/package.json:16]: Please use fixed dependencies (without `^`). Also, semaphore has realized a stable version, you can use 4.0.1 now.

**Review by cedoor — CHANGES_REQUESTED** [review]

File: packages/contracts/contracts/interfaces/ISemaphoreVoting.sol [hunk 40]
```diff
@@ -0,0 +1,77 @@
...
+    enum PollState {
+        Created,
+        Ongoing,
+        Ended
+    }
+
+    struct Poll {
+        address coordinator;
+        PollState state;
+        mapping(uint256 => bool) nullifiers;
```
Comment (cedoor) [inline packages/contracts/contracts/interfaces/ISemaphoreVoting.sol:21]: We can remove this, and the `groupId` parameter as well.

**Review by cedoor — CHANGES_REQUESTED** [review]

File: packages/contracts/contracts/interfaces/ISemaphoreVoting.sol (hunk 40 above)
Comment (Dozie2001) [reply packages/contracts/contracts/interfaces/ISemaphoreVoting.sol:21]: you mean nullifiers?

Comment (cedoor) [reply packages/contracts/contracts/interfaces/ISemaphoreVoting.sol:21]: Yess, they are already used on `Semaphore.sol`.

**Review by cedoor — APPROVED** [review]
There are still some issues with config files but we'll fix them after merging. I think the PR is ready.

Thanks a lot @Dozie2001 🙏🏽

**Pull request merged**